# Proyek Analisis Data: Bike Sharing Dataset
- **Nama:** Fitri Fatma Dewi
- **Email:** fitriworkspace12@gmail.com
- **ID Dicoding:** MC004D5X1425

## Menentukan Pertanyaan Bisnis

- **Bagaimana pengaruh kondisi cuaca terhadap jumlah penyewaan sepeda?**

Analisis ini bertujuan untuk memahami apakah faktor cuaca, seperti suhu, kelembaban, dan kondisi atmosfer, memiliki dampak signifikan terhadap jumlah penyewaan sepeda.
- **Kapan waktu optimal bagi pelanggan untuk menyewa sepeda berdasarkan tren penggunaan?**

Analisis ini akan mengidentifikasi pola penggunaan sepeda berdasarkan jam operasional untuk menentukan waktu dengan permintaan tertinggi dan terendah, sehingga dapat membantu optimalisasi layanan.


## Import Semua Packages/Library yang Digunakan

In [22]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

## Data Wrangling

### Gathering Data

In [23]:
# Data Day
day_df = pd.read_csv('Data/day.csv')
day_df.head()

instant      dteday  season  yr  mnth  holiday  weekday  workingday  \
0        1  2011-01-01       1   0     1        0        6           0   
1        2  2011-01-02       1   0     1        0        0           0   
2        3  2011-01-03       1   0     1        0        1           1   
3        4  2011-01-04       1   0     1        0        2           1   
4        5  2011-01-05       1   0     1        0        3           1   

   weathersit      temp     atemp       hum  windspeed  casual  registered  \
0           2  0.344167  0.363625  0.805833   0.160446     331         654   
1           2  0.363478  0.353739  0.696087   0.248539     131         670   
2           1  0.196364  0.189405  0.437273   0.248309     120        1229   
3           1  0.200000  0.212122  0.590435   0.160296     108        1454   
4           1  0.226957  0.229270  0.436957   0.186900      82        1518   

    cnt  
0   985  
1   801  
2  1349  
3  1562  
4  1600

In [24]:
# Data Hour
hour_df = pd.read_csv('Data/hour.csv')
hour_df.head()

instant      dteday  season  yr  mnth  hr  holiday  weekday  workingday  \
0        1  2011-01-01       1   0     1   0        0        6           0   
1        2  2011-01-01       1   0     1   1        0        6           0   
2        3  2011-01-01       1   0     1   2        0        6           0   
3        4  2011-01-01       1   0     1   3        0        6           0   
4        5  2011-01-01       1   0     1   4        0        6           0   

   weathersit  temp   atemp   hum  windspeed  casual  registered  cnt  
0           1  0.24  0.2879  0.81        0.0       3          13   16  
1           1  0.22  0.2727  0.80        0.0       8          32   40  
2           1  0.22  0.2727  0.80        0.0       5          27   32  
3           1  0.24  0.2879  0.75        0.0       3          10   13  
4           1  0.24  0.2879  0.75        0.0       0           1    1

**Insight:**
- Mengimport dan menampilkan tabel day
- Mingimport dan menampilkan tabel hour

### Assessing Data

##### Menilai Data day_df

In [25]:
day_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 731 entries, 0 to 730
Data columns (total 16 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   instant     731 non-null    int64  
 1   dteday      731 non-null    object 
 2   season      731 non-null    int64  
 3   yr          731 non-null    int64  
 4   mnth        731 non-null    int64  
 5   holiday     731 non-null    int64  
 6   weekday     731 non-null    int64  
 7   workingday  731 non-null    int64  
 8   weathersit  731 non-null    int64  
 9   temp        731 non-null    float64
 10  atemp       731 non-null    float64
 11  hum         731 non-null    float64
 12  windspeed   731 non-null    float64
 13  casual      731 non-null    int64  
 14  registered  731 non-null    int64  
 15  cnt         731 non-null    int64  
dtypes: float64(4), int64(11), object(1)
memory usage: 91.5+ KB


In [26]:
print("Jumlah duplikasi Data day_df: ", day_df.duplicated().sum())

Jumlah duplikasi Data day_df:  0


In [27]:
day_df.describe()

instant      season          yr        mnth     holiday     weekday  \
count  731.000000  731.000000  731.000000  731.000000  731.000000  731.000000   
mean   366.000000    2.496580    0.500684    6.519836    0.028728    2.997264   
std    211.165812    1.110807    0.500342    3.451913    0.167155    2.004787   
min      1.000000    1.000000    0.000000    1.000000    0.000000    0.000000   
25%    183.500000    2.000000    0.000000    4.000000    0.000000    1.000000   
50%    366.000000    3.000000    1.000000    7.000000    0.000000    3.000000   
75%    548.500000    3.000000    1.000000   10.000000    0.000000    5.000000   
max    731.000000    4.000000    1.000000   12.000000    1.000000    6.000000   

       workingday  weathersit        temp       atemp         hum   windspeed  \
count  731.000000  731.000000  731.000000  731.000000  731.000000  731.000000   
mean     0.683995    1.395349    0.495385    0.474354    0.627894    0.190486   
std      0.465233    0.544894    0.183051    0.162961    0.142429    0.077498   
min      0.000000    1.000000    0.059130    0.079070    0.000000    0.022392   
25%      0.000000    1.000000    0.337083    0.337842    0.520000    0.134950   
50%      1.000000    1.000000    0.498333    0.486733    0.626667    0.180975   
75%      1.000000    2.000000    0.655417    0.608602    0.730209    0.233214   
max      1.000000    3.000000    0.861667    0.840896    0.972500    0.507463   

            casual   registered          cnt  
count   731.000000   731.000000   731.000000  
mean    848.176471  3656.172367  4504.348837  
std     686.622488  1560.256377  1937.211452  
min       2.000000    20.000000    22.000000  
25%     315.500000  2497.000000  3152.000000  
50%     713.000000  3662.000000  4548.000000  
75%    1096.000000  4776.500000  5956.000000  
max    3410.000000  6946.000000  8714.000000

##### Menilai Data hour_df

In [28]:
hour_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17379 entries, 0 to 17378
Data columns (total 17 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   instant     17379 non-null  int64  
 1   dteday      17379 non-null  object 
 2   season      17379 non-null  int64  
 3   yr          17379 non-null  int64  
 4   mnth        17379 non-null  int64  
 5   hr          17379 non-null  int64  
 6   holiday     17379 non-null  int64  
 7   weekday     17379 non-null  int64  
 8   workingday  17379 non-null  int64  
 9   weathersit  17379 non-null  int64  
 10  temp        17379 non-null  float64
 11  atemp       17379 non-null  float64
 12  hum         17379 non-null  float64
 13  windspeed   17379 non-null  float64
 14  casual      17379 non-null  int64  
 15  registered  17379 non-null  int64  
 16  cnt         17379 non-null  int64  
dtypes: float64(4), int64(12), object(1)
memory usage: 2.3+ MB


In [29]:

print("Jumlah duplikasi Data hour_df: ", hour_df.duplicated().sum())

Jumlah duplikasi Data hour_df:  0


In [30]:
hour_df.describe()

instant        season            yr          mnth            hr  \
count  17379.0000  17379.000000  17379.000000  17379.000000  17379.000000   
mean    8690.0000      2.501640      0.502561      6.537775     11.546752   
std     5017.0295      1.106918      0.500008      3.438776      6.914405   
min        1.0000      1.000000      0.000000      1.000000      0.000000   
25%     4345.5000      2.000000      0.000000      4.000000      6.000000   
50%     8690.0000      3.000000      1.000000      7.000000     12.000000   
75%    13034.5000      3.000000      1.000000     10.000000     18.000000   
max    17379.0000      4.000000      1.000000     12.000000     23.000000   

            holiday       weekday    workingday    weathersit          temp  \
count  17379.000000  17379.000000  17379.000000  17379.000000  17379.000000   
mean       0.028770      3.003683      0.682721      1.425283      0.496987   
std        0.167165      2.005771      0.465431      0.639357      0.192556   
min        0.000000      0.000000      0.000000      1.000000      0.020000   
25%        0.000000      1.000000      0.000000      1.000000      0.340000   
50%        0.000000      3.000000      1.000000      1.000000      0.500000   
75%        0.000000      5.000000      1.000000      2.000000      0.660000   
max        1.000000      6.000000      1.000000      4.000000      1.000000   

              atemp           hum     windspeed        casual    registered  \
count  17379.000000  17379.000000  17379.000000  17379.000000  17379.000000   
mean       0.475775      0.627229      0.190098     35.676218    153.786869   
std        0.171850      0.192930      0.122340     49.305030    151.357286   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.333300      0.480000      0.104500      4.000000     34.000000   
50%        0.484800      0.630000      0.194000     17.000000    115.000000   
75%        0.621200      0.780000      0.253700     48.000000    220.000000   
max        1.000000      1.000000      0.850700    367.000000    886.000000   

                cnt  
count  17379.000000  
mean     189.463088  
std      181.387599  
min        1.000000  
25%       40.000000  
50%      142.000000  
75%      281.000000  
max      977.000000

**Insight:**
- Untuk Data **day_df**, masing-masing variabelnya memiliki jumlah yang sama, sehingga tidak terdapat missing value.
- Uji Duplikat data pada **day_df** juga menunjukkan jumlah 0, artinya tidak terdapat data duplikat.
- Pada Data **hour_df**, masing-masing variabelnya memiliki jumlah yang sama, sehingga tidak terdapat missing value.
- Dilakukan Uji Duplikat data pula untuk **hour_df**, dan terindikasi tidak terdapat data duplikat.
- Namun, variabel dteday pada kedua data di atas, terindikasi terdapat kesalahan type data, harusnya ia bertipe data "date", bukan "object". **Akan di tangani pada tahap data cleaning.**
- Terakhir, untuk analisis deskripsi statistik, kedua data terindikasi aman, dan tidak ada nilai data yang bermasalah. Semuanya nilai data sudah sesuai.

### Cleaning Data

##### Mengganti type data variabel dteday

In [31]:
datetime_columns = ["dteday"]
 
for column in datetime_columns:
  day_df[column] = pd.to_datetime(day_df[column])
  hour_df[column] = pd.to_datetime(hour_df[column])

In [32]:
day_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 731 entries, 0 to 730
Data columns (total 16 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   instant     731 non-null    int64         
 1   dteday      731 non-null    datetime64[ns]
 2   season      731 non-null    int64         
 3   yr          731 non-null    int64         
 4   mnth        731 non-null    int64         
 5   holiday     731 non-null    int64         
 6   weekday     731 non-null    int64         
 7   workingday  731 non-null    int64         
 8   weathersit  731 non-null    int64         
 9   temp        731 non-null    float64       
 10  atemp       731 non-null    float64       
 11  hum         731 non-null    float64       
 12  windspeed   731 non-null    float64       
 13  casual      731 non-null    int64         
 14  registered  731 non-null    int64         
 15  cnt         731 non-null    int64         
dtypes: datetime64[ns](1), floa

In [33]:
hour_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17379 entries, 0 to 17378
Data columns (total 17 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   instant     17379 non-null  int64         
 1   dteday      17379 non-null  datetime64[ns]
 2   season      17379 non-null  int64         
 3   yr          17379 non-null  int64         
 4   mnth        17379 non-null  int64         
 5   hr          17379 non-null  int64         
 6   holiday     17379 non-null  int64         
 7   weekday     17379 non-null  int64         
 8   workingday  17379 non-null  int64         
 9   weathersit  17379 non-null  int64         
 10  temp        17379 non-null  float64       
 11  atemp       17379 non-null  float64       
 12  hum         17379 non-null  float64       
 13  windspeed   17379 non-null  float64       
 14  casual      17379 non-null  int64         
 15  registered  17379 non-null  int64         
 16  cnt         17379 non-

##### Mengganti nama variabel yang susah diingat, agar lebih mudah diingat

In [34]:
day_df.rename(columns={
    "dteday": "date",
    "yr": "year",
    "mnth": "month",
    "hum":"humidity",
    "cnt":"count"
}, inplace=True)

In [35]:
day_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 731 entries, 0 to 730
Data columns (total 16 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   instant     731 non-null    int64         
 1   date        731 non-null    datetime64[ns]
 2   season      731 non-null    int64         
 3   year        731 non-null    int64         
 4   month       731 non-null    int64         
 5   holiday     731 non-null    int64         
 6   weekday     731 non-null    int64         
 7   workingday  731 non-null    int64         
 8   weathersit  731 non-null    int64         
 9   temp        731 non-null    float64       
 10  atemp       731 non-null    float64       
 11  humidity    731 non-null    float64       
 12  windspeed   731 non-null    float64       
 13  casual      731 non-null    int64         
 14  registered  731 non-null    int64         
 15  count       731 non-null    int64         
dtypes: datetime64[ns](1), floa

In [36]:
hour_df.rename(columns={
    "dteday": "date",
    "yr": "year",
    "mnth": "month",
    "hum":"humidity",
    "hr": "hour",
    "cnt":"count"
}, inplace=True)

In [37]:
hour_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17379 entries, 0 to 17378
Data columns (total 17 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   instant     17379 non-null  int64         
 1   date        17379 non-null  datetime64[ns]
 2   season      17379 non-null  int64         
 3   year        17379 non-null  int64         
 4   month       17379 non-null  int64         
 5   hour        17379 non-null  int64         
 6   holiday     17379 non-null  int64         
 7   weekday     17379 non-null  int64         
 8   workingday  17379 non-null  int64         
 9   weathersit  17379 non-null  int64         
 10  temp        17379 non-null  float64       
 11  atemp       17379 non-null  float64       
 12  humidity    17379 non-null  float64       
 13  windspeed   17379 non-null  float64       
 14  casual      17379 non-null  int64         
 15  registered  17379 non-null  int64         
 16  count       17379 non-

**Insight:**
- Mengganti type variabel dtaday, dari  "object" menjadi "datetime"
- Mengganti nama variabel yang susah ditulis/diingat seperti **dteday, yr, mnth, hum, hr, dan cnt**. Menjadi **date,year, month, humidity, hour, dan count.**

## Exploratory Data Analysis (EDA)

##### Exploratory Data day_df

In [38]:
day_df.describe(include="all")

instant                 date      season        year       month  \
count  731.000000                  731  731.000000  731.000000  731.000000   
mean   366.000000  2012-01-01 00:00:00    2.496580    0.500684    6.519836   
min      1.000000  2011-01-01 00:00:00    1.000000    0.000000    1.000000   
25%    183.500000  2011-07-02 12:00:00    2.000000    0.000000    4.000000   
50%    366.000000  2012-01-01 00:00:00    3.000000    1.000000    7.000000   
75%    548.500000  2012-07-01 12:00:00    3.000000    1.000000   10.000000   
max    731.000000  2012-12-31 00:00:00    4.000000    1.000000   12.000000   
std    211.165812                  NaN    1.110807    0.500342    3.451913   

          holiday     weekday  workingday  weathersit        temp       atemp  \
count  731.000000  731.000000  731.000000  731.000000  731.000000  731.000000   
mean     0.028728    2.997264    0.683995    1.395349    0.495385    0.474354   
min      0.000000    0.000000    0.000000    1.000000    0.059130    0.079070   
25%      0.000000    1.000000    0.000000    1.000000    0.337083    0.337842   
50%      0.000000    3.000000    1.000000    1.000000    0.498333    0.486733   
75%      0.000000    5.000000    1.000000    2.000000    0.655417    0.608602   
max      1.000000    6.000000    1.000000    3.000000    0.861667    0.840896   
std      0.167155    2.004787    0.465233    0.544894    0.183051    0.162961   

         humidity   windspeed       casual   registered        count  
count  731.000000  731.000000   731.000000   731.000000   731.000000  
mean     0.627894    0.190486   848.176471  3656.172367  4504.348837  
min      0.000000    0.022392     2.000000    20.000000    22.000000  
25%      0.520000    0.134950   315.500000  2497.000000  3152.000000  
50%      0.626667    0.180975   713.000000  3662.000000  4548.000000  
75%      0.730209    0.233214  1096.000000  4776.500000  5956.000000  
max      0.972500    0.507463  3410.000000  6946.000000  8714.000000  
std      0.142429    0.077498   686.622488  1560.256377  1937.211452

In [45]:
day_df.groupby(by="weathersit").agg({
    "instant": "nunique",
    "count": ["max", "min", "mean", "std"]
})

instant count                               
           nunique   max  min         mean          std
weathersit                                             
1              463  8714  431  4876.786177  1879.483989
2              247  8362  605  4035.862348  1809.109918
3               21  4639   22  1803.285714  1240.284449

Berdasarkan pivot table di atas, dapat disimpulkan bahwa penyewa yang paling banyak terdapat pada saat kondisi cuaca pada kategori 1, yakni saat cuaca *Clear* (Cerah), *Few clouds* (sedikit berawan), *Partly cloudy* (sebagian berawan). Namun keragaman banyaknya penyewa pada kondisi cuaca 1 juga paling besar, ditunjukkan dengan standard deviasi terbesar, artinya jumlah penyewa pada kondisi cuaca 1 sangat beragam dengan minimal 431 penyewa dan maksimal 463 penyewa.

In [43]:
day_df.groupby(by="season").instant.nunique().sort_values(ascending=False)

season
3    188
2    184
1    181
4    178
Name: instant, dtype: int64

##### Exploratory Data hour_df

In [46]:
hour_df.describe()

instant                           date        season          year  \
count  17379.0000                          17379  17379.000000  17379.000000   
mean    8690.0000  2012-01-02 04:08:34.552045568      2.501640      0.502561   
min        1.0000            2011-01-01 00:00:00      1.000000      0.000000   
25%     4345.5000            2011-07-04 00:00:00      2.000000      0.000000   
50%     8690.0000            2012-01-02 00:00:00      3.000000      1.000000   
75%    13034.5000            2012-07-02 00:00:00      3.000000      1.000000   
max    17379.0000            2012-12-31 00:00:00      4.000000      1.000000   
std     5017.0295                            NaN      1.106918      0.500008   

              month          hour       holiday       weekday    workingday  \
count  17379.000000  17379.000000  17379.000000  17379.000000  17379.000000   
mean       6.537775     11.546752      0.028770      3.003683      0.682721   
min        1.000000      0.000000      0.000000      0.000000      0.000000   
25%        4.000000      6.000000      0.000000      1.000000      0.000000   
50%        7.000000     12.000000      0.000000      3.000000      1.000000   
75%       10.000000     18.000000      0.000000      5.000000      1.000000   
max       12.000000     23.000000      1.000000      6.000000      1.000000   
std        3.438776      6.914405      0.167165      2.005771      0.465431   

         weathersit          temp         atemp      humidity     windspeed  \
count  17379.000000  17379.000000  17379.000000  17379.000000  17379.000000   
mean       1.425283      0.496987      0.475775      0.627229      0.190098   
min        1.000000      0.020000      0.000000      0.000000      0.000000   
25%        1.000000      0.340000      0.333300      0.480000      0.104500   
50%        1.000000      0.500000      0.484800      0.630000      0.194000   
75%        2.000000      0.660000      0.621200      0.780000      0.253700   
max        4.000000      1.000000      1.000000      1.000000      0.850700   
std        0.639357      0.192556      0.171850      0.192930      0.122340   

             casual    registered         count  
count  17379.000000  17379.000000  17379.000000  
mean      35.676218    153.786869    189.463088  
min        0.000000      0.000000      1.000000  
25%        4.000000     34.000000     40.000000  
50%       17.000000    115.000000    142.000000  
75%       48.000000    220.000000    281.000000  
max      367.000000    886.000000    977.000000  
std       49.305030    151.357286    181.387599

In [61]:
def classify_time(hour):
    if 0 <= hour < 5:
        return "Dini Hari"   
    elif 5 <= hour < 10:
        return "Pagi"
    elif 10 <= hour < 15:
        return "Siang"
    elif 15 <= hour < 18:
        return "Sore"
    else:
        return "Malam"

hour_df["time_category"] = hour_df["hour"].apply(classify_time)
hour_df.sample(5)

instant       date  season  year  month  hour  holiday  weekday  \
9975      9976 2012-02-25       1     1      2    16        0        6   
16091    16092 2012-11-08       4     1     11     5        0        4   
7739      7740 2011-11-24       4     0     11     2        1        4   
6298      6299 2011-09-25       4     0      9     0        0        0   
13456    13457 2012-07-19       3     1      7    21        0        4   

       workingday  weathersit  temp   atemp  humidity  windspeed  casual  \
9975            0           1  0.32  0.2727      0.26     0.5821      32   
16091           1           2  0.30  0.2879      0.42     0.2239       0   
7739            0           1  0.30  0.2879      0.61     0.2836       3   
6298            0           2  0.60  0.5455      0.88     0.0000      37   
13456           1           3  0.66  0.5909      0.89     0.2537      38   

       registered  count time_category  
9975          163    195          Sore  
16091          35     35          Pagi  
7739           19     22     Dini Hari  
6298          136    173     Dini Hari  
13456         294    332         Malam

In [62]:
hour_df.groupby(by="time_category").instant.nunique().sort_values(ascending=False)

time_category
Malam        4368
Siang        3640
Pagi         3623
Dini Hari    3559
Sore         2189
Name: instant, dtype: int64